Modules/_tracemalloc.c

```diff
@@ -78,15 +78,21 @@ __attribute__((packed))
 
 typedef struct {
     Py_uhash_t hash;
-    int nframe;
+    /* Number of frames stored */
+    uint16_t nframe;
+    /* Total number of frames the traceback had */
+    uint16_t total_nframe;
     frame_t frames[1];
 } traceback_t;
 
 #define TRACEBACK_SIZE(NFRAME) \
         (sizeof(traceback_t) + sizeof(frame_t) * (NFRAME - 1))
 
+/* The maximum number of frames is either:
+ - The maximum number of frames we can store in `nframe`
+ - The maximum memory size_t we can allocate */
 #define MAX_NFRAME \
-        ((INT_MAX - (int)sizeof(traceback_t)) / (int)sizeof(frame_t) + 1)
+        Py_MIN(UINT16_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1))
```

Comment:
Can you try to convert to macro to a constant variable? Otherwise, the macro is expended each time it is used, and the compiler might recompute it each time.

```
static const int MAX_NFRAME = Py_MIN(UINT16_MAX, ((SIZE_MAX - sizeof(traceback_t)) / sizeof(frame_t) + 1));
```

I suggest to use int (signed). If you want to use an unsigned int, you need to modify more change.

You may need to update code around `if (max_nframe < 1 || (unsigned long) max_nframe > MAX_NFRAME) {`.

Reply:
Not sure it's worth it, but fine. AFAIU the macro is expanded only when pre-processing `_tracemalloc.` which should only happen once.

The result is an `unsigned long`, no matter what. We can cast it to something else, but that should be ok — the code compiles without warning AFAICT.